### PR TITLE
Codemod: Convert `.stories.mdx` to MDX and CSF

### PR DESCRIPTION
**Pull request by kasperpeulen**
Issues: https://github.com/storybookjs/storybook/issues/19497 https://github.com/storybookjs/storybook/issues/20733

## What I did

This should convert mdx files to mdx and csf. 
Please check the test file mainly, the implementation is quite hairy.

### CODE REVIEW

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (tmeasday) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:27]: Is it weird that it adds this line when no stories were defined in the file?

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (tmeasday) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:91]: Did we decide that resolving `id`s was too tricky?

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (tmeasday) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:113]: Should we check the CSF in this test?

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (tmeasday) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:229]: Template is undefined here (but was undefined in the MDX file too). Are we only bringing variable references across if they actually exist?

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (tmeasday) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:343]: Should this file `import React from 'react'`?

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (tmeasday) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:380]: Can we have a test for "story child is CSF3"?

**Review by tmeasday — CHANGES_REQUESTED** [review]
This is great @kasperpeulen 👏 

Some comments/questions. I didn't really read the transform code I'll leave that to @shilman etc.

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (JReinhold) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:91]: If so, would it make sense to inject a comment explaining that the `id` is deprecated?
like
```
{/* <Story id="button--primary" /> is deprecated, please migrate it to <Story of={referenceToStory} /> manually */}
```

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (JReinhold) [inline code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:288]: Same undefined Template here

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:27]: Right, changed that.

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:112]: ✅

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts]: Good point, I made the test correct. We don't check if references actually exist, but we do make sure to copy all ESM code from the mdx file to the CSF file, so all (and maybe more) references should exist.

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:367]: From React 17 it is not needed anymore, and also back-ported to the latest 16 release. 
There are ESLint rules to automatically fix this if you need it, so I kind of feel better with leaving it out of this codemod.

**Review by JReinhold — COMMENTED** [review]
I don't have much to add besides @tmeasday comments. looks good!

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (JReinhold) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts]: Maybe I was unclear, I think we should add a comment _above_ the Story, not wrap the whole Story in a comment. That way it would still work, but we would explain the user what they need to do.

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:404]: Added it

File: code/lib/codemod/src/transforms/mdx-to-csf.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/mdx-to-csf.ts]: Thanks, done!

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts:309]: Fixed it

**Review by tmeasday — APPROVED** [review]
LGTM! But I didn't look to closely at the AST code.

File: code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts]: Got it, fixed it, hopefully my tree traversing logic for skipping the mutated entries is fail proof 🤞

**Review by ndelangen — APPROVED** [review]

**Comment by shilman** [comment]
@kasperpeulen This is really amazing work 💪 

Did a quick test on https://github.com/RocketCommunicationsInc/astro/tree/main/packages/web-components and found the following issues. Your call on what to fix before merging vs what to fix during follow-up vs what to ignore as a corner case. Filing each issue as a separate comment and you can create an issue from the comment as needed by right-clicking on the comment and "Reference in new issue"

**Comment by shilman** [comment]
On several files there is a syntax error in babel. Th error is non-obvious on first inspection -- could be a user error?

Example: https://raw.githubusercontent.com/RocketCommunicationsInc/astro/main/packages/web-components/src/stories/tree.stories.mdx

```
 ERR packages/web-components/src/stories/tree.stories.mdx Transformation error (Unexpected token, expected "," (1:8))
SyntaxError: Unexpected token, expected "," (1:8)
    at instantiate (/Users/shilman/projects/storybookjs/storybook/code/node_modules/@babel/parser/lib/index.js:64:32)
    at constructor (/Users/shilman/projects/storybookjs/storybook/code/node_modules/@babel/parser/lib/index.js:362:12)
```

Furthermore, despite the error, the file is renamed from `tree.stories.mdx` to `tree.mdx`. This is a bug in the codemod logic.

**Comment by shilman** [comment]
In some stories, there is a duplicate variable name in the output.

Example: https://raw.githubusercontent.com/RocketCommunicationsInc/astro/main/packages/web-components/src/stories/accordion-item.stories.mdx

This generates:
```js
```

**Comment by shilman** [comment]
Here's another issue from the same repo:

https://raw.githubusercontent.com/RocketCommunicationsInc/astro/main/packages/web-components/src/stories/astro-uxds/welcome/StartHere.stories.mdx

Generates:

```mdx
// StartHere.mdx
import { Meta } from '@storybook/blocks'
import * as StartHereStories from './StartHere.stories';

<Meta of={StartHereStories} />

# Welcome to the Astro Storybook
...
```

```js
// StartHere.stories.js
export default {
    title: 'Astro UXDS/Welcome/Start Here',
}
```

Because the story file has no exports, the storybook fails to load. I believe there is another automigration `bare-mdx-stories-glob` that tries to handle this, but for some reason it didn't run on this repo when I upgraded (perhaps an unrelated bug?). Nevertheless, maybe this codemod should be robust to that situation anyway?

**Comment by shilman** [comment]
Here's another one from

https://raw.githubusercontent.com/RocketCommunicationsInc/astro/main/packages/web-components/src/stories/global-status-bar.stories.mdx

```
// global-status-bar.mdx
 import * as global-status-barStories from './global-status-bar.stories';

<Meta of={global-status-barStories} />
```

**Review by shilman — APPROVED** [review]
This is amazing @kasperpeulen -- great job!!!

**Pull request merged**
